tree/tree.go

```diff
@@ -0,0 +1,723 @@
+package tree
+
+import (
+	"fmt"
+	"os"
+
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/charmbracelet/lipgloss"
+	ltree "github.com/charmbracelet/lipgloss/tree"
+
+	"github.com/charmbracelet/bubbles/help"
+	"github.com/charmbracelet/bubbles/key"
+)
+
+// StyleFunc allows the tree to be styled per item.
+type StyleFunc func(children Nodes, i int) lipgloss.Style
+
+// Styles contains style definitions for this tree component. By default, these
+// values are generated by DefaultStyles.
+type Styles struct {
+	SelectionCursor       lipgloss.Style
+	HelpStyle             lipgloss.Style
+	TreeStyle             lipgloss.Style
+	selectedNodeFunc      StyleFunc
+	SelectedNodeStyle     lipgloss.Style
+	SelectedNodeStyleFunc StyleFunc
+	nodeFunc              StyleFunc
+	NodeStyle             lipgloss.Style
+	NodeStyleFunc         StyleFunc
+}
+
+// DefaultStyles returns a set of default style definitions for this tree
+// component.
+func DefaultStyles() (s Styles) {
+	s.NodeStyle = lipgloss.NewStyle()
+	s.SelectedNodeStyle = lipgloss.NewStyle().
+		Background(lipgloss.Color("62")).
+		Foreground(lipgloss.Color("18")).
+		Bold(true)
+	s.nodeFunc = func(_ Nodes, _ int) lipgloss.Style {
+		return s.NodeStyle
+	}
+	s.selectedNodeFunc = func(_ Nodes, _ int) lipgloss.Style {
+		return s.SelectedNodeStyle
+	}
+	s.SelectionCursor = lipgloss.NewStyle().Foreground(lipgloss.Color("9"))
+	s.HelpStyle = lipgloss.NewStyle().PaddingTop(1)
+
+	return s
+}
+
+// KeyMap is the key bindings for different actions within the tree.
+type KeyMap struct {
+	Down   key.Binding
+	Up     key.Binding
+	Toggle key.Binding
+
+	// Help toggle keybindings.
+	ShowFullHelp  key.Binding
+	CloseFullHelp key.Binding
+
+	Quit key.Binding
+}
+
+// DefaultKeyMap is the default set of key bindings for navigating and acting
+// upon the tree.
+var DefaultKeyMap = KeyMap{
```

Comment:
probably need to add more keybinds, I was think also about:
- half a page up / down which would equal 10 rows let's say?
- anything else?

Reply:
I think generally we can probably mirror viewport keybindings here.

Also, one thing we're doing in `v2` is removing globals when possible. So I'd change the above to a function:

```go
func DefaultKeyMap() KeyMap
```